src/renderer/qss/js/qssFocusManager.js

```diff
@@ -0,0 +1,331 @@
+ * compliance with this License.
+ * The research leading to these results has received funding from the European Union's
+ * Seventh Framework Programme (FP7/2007-2013) under grant agreement no. 289016.
+ * You may obtain a copy of the License at
+ * https://github.com/GPII/universal/blob/master/LICENSE.txt
+ */
+
+/* global fluid jQuery */
+
+"use strict";
+(function (fluid, jQuery) {
+    var gpii = fluid.registerNamespace("gpii");
+
+    /**
+     * A special type of focusManager which enables navigation between the focusable
+     * elements using Tab, Shift + Tab, Arrow Up, Arrow Down, Arrow Left and the
+     * Arrow Right key. The navigation focus basically moves in the visual direction.
+     */
+    fluid.defaults("gpii.qss.qssFocusManager", {
+        gradeNames: ["gpii.qss.horizontalFocusManager", "gpii.qss.verticalFocusManager"],
+
+        maxElementsPerFocusGroup: 2,
+
+        styles: {
+            button: "fl-qss-button",
+            smallButton: "fl-qss-smallButton"
+        },
+
+        listeners: {
+            "onArrowUpPressed.impl": {
+                funcName: "gpii.qss.qssFocusManager.onArrowUpPressed",
+                args: ["{that}"]
+            },
+            "onArrowDownPressed.impl": {
+                funcName: "gpii.qss.qssFocusManager.onArrowDownPressed",
+                args: ["{that}"]
+            },
+            "onArrowLeftPressed.impl": {
+                funcName: "gpii.qss.qssFocusManager.onArrowLeftPressed",
+                args: ["{that}"]
+            },
+            "onArrowRightPressed.impl": {
+                funcName: "gpii.qss.qssFocusManager.onArrowRightPressed",
+                args: ["{that}"]
+            }
+        },
+
+        invokers: {
+            getFocusInfo: {
+                funcName: "gpii.qss.qssFocusManager.getFocusInfo",
+                args: ["{that}.container", "{that}.options.styles"]
+            },
+            getFocusGroupsInfo: {
+                funcName: "gpii.qss.qssFocusManager.getFocusGroupsInfo",
+                args: ["{that}", "{that}.container"]
+            }
+        }
+    });
+
+    /**
+     * Retrieves the tab index of a given DOM or jQuery element as a number.
+     * @param {HTMLElement | jQuery} element - A simple DOM element or wrapped in a jQuery
+     * object.
+     * @return {Number} The tab index of the given element or `undefined` if this property
+     * is not specified.
+     */
+    gpii.qss.qssFocusManager.getTabIndex = function (element) {
+        var tabIndex = jQuery(element).attr("tabindex");
+        if (tabIndex) {
+            return parseInt(tabIndex, 10);
+        }
+    };
+
+    /**
+     * Returns information about the focusable elements in the QSS as well as the index of
+     * the currently focused element. "Focusable element" in terms of the QSS means an element
+     * with a "fl-focusable" class which has a specified `tabindex` greater than or equal to 0.
```

Comment:
It might be useful if this is added to the description of the component

Reply:
I was rather thinking of a bit more detailed documentation describing the component behaviour and usage. I think it would be useful to have details on the whole process that resides at one place.

Reply:
I am not sure about that. I think it may be more confusing to describe everything in one place given that the logic is totally different in the two scenarios.